From a3cc9361b6b04333df885b9b0fa59f2ecd0f705f Mon Sep 17 00:00:00 2001
From: Matt Fleming <matt.fleming@intel.com>
Date: Fri, 4 Oct 2013 09:36:56 +0100
Subject: [PATCH 1/3] x86/efi: Add EFI framebuffer earlyprintk support
References: FATE#317449
Patch-Mainline: v3.15

It's incredibly difficult to diagnose early EFI boot issues without
special hardware because earlyprintk=vga doesn't work on EFI systems.

Add support for writing to the EFI framebuffer, via earlyprintk=efi,
which will actually give users a chance of providing debug output.

oneukum: Edited for we don't do mixed 32/64 environments

Cc: H. Peter Anvin <hpa@zytor.com>
Acked-by: Ingo Molnar <mingo@kernel.org>
Cc: Thomas Gleixner <tglx@linutronix.de>
Cc: Peter Jones <pjones@redhat.com>
Signed-off-by: Matt Fleming <matt.fleming@intel.com>
Signed-off-by: Oliver Neukum <oneukum@suse.de>

Automatically created from "patches.drivers/0001-x86-efi-Add-EFI-framebuffer-earlyprintk-support.patch" by xen-port-patches.py

--- sle11sp4.orig/arch/x86/Kconfig.debug	2013-12-02 17:50:19.000000000 +0100
+++ sle11sp4/arch/x86/Kconfig.debug	2013-12-06 15:28:36.000000000 +0100
@@ -69,7 +69,7 @@ config EARLY_PRINTK_DBGP
 
 config EARLY_PRINTK_EFI
 	bool "Early printk via the EFI framebuffer"
-	depends on EFI && EARLY_PRINTK
+	depends on EFI && EARLY_PRINTK && !XEN
 	select FONT_SUPPORT
 	---help---
 	  Write kernel log output directly into the EFI framebuffer.
--- sle11sp4.orig/arch/x86/kernel/early_printk-xen.c	2011-02-01 15:41:35.000000000 +0100
+++ sle11sp4/arch/x86/kernel/early_printk-xen.c	2015-02-02 15:51:22.000000000 +0100
@@ -16,6 +16,7 @@
 #include <asm/mrst.h>
 #include <asm/pgtable.h>
 #include <linux/usb/ehci_def.h>
+#include <linux/efi.h>
 
 #ifndef CONFIG_XEN
 /* Simple VGA output */
@@ -260,7 +261,8 @@ static int __init setup_early_printk(cha
 			max_ypos = boot_params.screen_info.orig_video_lines;
 			current_ypos = boot_params.screen_info.orig_y;
 #else
-		if (!strncmp(buf, "vga", 3) || !strncmp(buf, "xen", 3)) {
+		if (!strncmp(buf, efi_enabled ? "efi" : "vga", 3)
+		    || !strncmp(buf, "xen", 3)) {
 #endif
 			early_console_register(&early_vga_console, keep);
 		}
@@ -283,6 +285,7 @@ static int __init setup_early_printk(cha
 			early_console_register(&early_hsu_console, keep);
 		}
 #endif
+
 		buf++;
 	}
 	return 0;
